### ISSUE TITLE
failing to compile on RPI3 

### ISSUE THREAD

**Issue by ghost** [ptitSeb/kam_remake#1]
Hi Ptit. so, he we are!

Ive set lazarus (2.0) to target linux arm (armv7) and it fails on dglOpenGL. there are some things related to fpc than I should check (it doesnt find a lot of things on /usr/lib/arm-linux-gnueabihf/fpc) but the crashing event was related to something called "Error: invalid offset, value too large" (many many times) when lazarus execute "/usr/bin/as" with command line "-mfloat-abi=hard -meabi=5 (btw, this -meabi=5 its a pandora thing??) -march=armv7 -mfpu=vfpv2 -o /home/pi/Desktop/kam_remake/dglOpenGL.o  /home/pi/Desktop/kam_remake/dglOpenGL.s"

thats what he says this long message.txt

[messages.txt](https://github.com/ptitSeb/kam_remake/files/3294985/messages.txt)

**Comment by ghost** [comment]
ptit, you should check this: http://forum.ragezone.com/f1012/release-city-heroes-city-villains-1161004/
I hate marvel world... but millions doesn't

**Comment by ptitSeb** [comment]
Thanks for the info, but it's a MMO, I'm not sure I want to invest any time in that. Interresting story anyway.

About your Lazarus issue. I don't know. It should be plug'n play now, so make your your lazarus install is working fine, by compiling some simple example first maybe.

**Comment by ghost** [comment]
Ok, I will try that! Thanks anyway, a friend of mine on california love it. I dont like them too..the only mmo ive played its open source from the beginning back in  2001.... I am sooo old hahaha

**Comment by Askmewho** [comment]
Ptit, could you be so mercyful on crosscompiling the binaries to armhf?? Pandora its armv6 ,does it??

**Comment by ghost** [comment]
@ptitSeb do you know any way to create a 32bit environment inside a 64bit os (on my rk3399) to run bastion (fmod its armhf only lib)... or there is no way to do that? a bit stupid by fmod makers to support armhf and not aarch64 imho

**Comment by ptitSeb** [comment]
Thre is probably a way to create a chroot 32bits environnement, or maybe just play with the muti-arch capability of debain. But I haven't tried yet to do that on my 64bits boards, so I haven't got any real hint on how to do it for now.

**Comment by ghost** [comment]
Thanks anyway.

**Comment by ghost** [comment]
Ive asked Aarch64 support on fmod forum (not as nicely it had should been), and they will prepare a build for next release :) 

i am trying a relatively new fna game , at least for me, its mercenary kings.. but ive get some weird output
```

rockpi@rockpi:~/Desktop/Mercenary.Kings.v1.5.0.21320$ LD_LIBRARY_PATH=/home/rockpi/Desktop/monolibs mono MercenaryKings.exe
[sudo] password for rockpi: 
AL lib: (WW) jack_client_open() failed, 0x01
AL lib: (WW) Failed to initialize backend "jack"
IGLDevice: OpenGLDevice
OpenGL Device: panfrost
OpenGL Driver: 2.1 Mesa 19.3.0-devel (git-ac5bda374a)
OpenGL Vendor: panfrost
MojoShader Profile: glsl120
EXT_swap_control_tear unsupported. Fall back to standard VSync.
Mesa: User error: GL_INVALID_ENUM in glTexParameter(pname=GL_TEXTURE_MAX_ANISOTROPY)
EXT_swap_control_tear unsupported. Fall back to standard VSync.
Final Guids
 - Guid 0 - 3b61b6ec-372e-4103-8681-44ff3e903689
 - Guid 1 - 8cec8827-245e-4a29-8e2e-795338f83d9b
 - Guid 2 - 557cab77-b1a3-48fc-80d8-dae658e23cef
 - Guid 3 - 8acc3d0f-8eeb-4926-a795-dbb25cdf4711
[ERROR] FATAL UNHANDLED EXCEPTION: Paris.Paris+VCRedistNotInstalledException: Visual C++ 2010 Redistributable Package is not installed.
  at Paris.Paris.InitPlatform () [0x00058] in <234f3c7891474a61abdadddf162c6a49>:0 
  at Paris.Paris.Init () [0x0006c] in <234f3c7891474a61abdadddf162c6a49>:0 
  at Paris.Paris.Tick (System.Single i_fDeltaTime) [0x00008] in <234f3c7891474a61abdadddf162c6a49>:0 
  at Paris.Paris.Update (Microsoft.Xna.Framework.GameTime i_gameTime) [0x0000f] in <234f3c7891474a61abdadddf162c6a49>:0 
  at Microsoft.Xna.Framework.Game.Tick () [0x00108] in <fc5ddff044aa42e1a0c88d1cb8c9f9a9>:0 
  at Microsoft.Xna.Framework.SDL2_FNAPlatform.RunLoop (Microsoft.Xna.Framework.Game game) [0x005d5] in <fc5ddff044aa42e1a0c88d1cb8c9f9a9>:0 
  at Microsoft.Xna.Framework.Game.Run () [0x0002c] in <fc5ddff044aa42e1a0c88d1cb8c9f9a9>:0 
  at Paris.Program.Main (System.String[] args) [0x0007d] in <234f3c7891474a61abdadddf162c6a49>:0 

```

the vs libs arent included on mono-complete & mono-devel? 

this board its a monster..ive get used to rpi3... but this can encode the output while gaming very easily.

**Comment by Askmewho** [comment]
@ptitSeb I have another problem with mono games, ive only tested owlboy, and I have a weird sound effects problem. I suppose its openal, but they are so many libs related to audio i dont know what could be. did you suffer the same issue on any game like this??
https://drive.google.com/file/d/1iLrUDE6ZQZP_aPmScrGE-XbAe3ZOlTPx/view?usp=sharing

**Comment by ptitSeb** [comment]
Mercenary Kings, I don't have the games. Strange error, I have no idea what it is testing to know if VC redist are there or not. I'll see if I can find some game with the same behavour.

As for Owlboy, I had no issue on my side with sound. Games worked fine.

**Comment by ghost** [comment]
Yeah, it should be pulse audio. It give me that kind of problems on the
past. I hate it haha. Thanks!

**Comment by ghost** [comment]
Hi @ptitSeb  !! stardew valley and hammerwatch worked without issue. that newer game, mercenary kings, havent an non-drm release, thats why it will not work, its claiming for steam stuff. hope they release a newer non-drm version because the game looks great. I coulnt make iconoclast run, did you tried it? seems to be a very good game. another thing I couldn't make work was dust an Elysian , maybe my version its not the latest (seems to not be updated to fna), which version did you use?? hope you are fine. 

I dont have time to test your kam_remake on my rockpi4 (surely it will not work), but it could work on arm64 one day?

**Comment by ptitSeb** [comment]
For Dust, it's a GoG version. IIRC.
Mercenary Kings: ah DRM, yeah, that could be an issue. I have Terraria like that, that I cannot run because it use Steam DRM, and I don't have a GoG or HB drm-free version.

Kam_remake on 64bits? I don't know. I still need to validate how it runs on 32bits first.

**Comment by ghost** [comment]
Thanks! The mono debug its really handy. Check my last video! https://youtu.be/UeAmHmIbIRI

As you know, I am testing panfrost, just because they promise opengl 3.x and even vulkan on the long road. Best wishes.

**Comment by ptitSeb** [comment]
I didn't really got what you said as an intro, but Stardew Valley seems to runs very well. For Hammerwatch,some  advanced rendering mode looks all black, and the record of the game if very choppy. I don't know if it's like that for real or if it's the recording, but this doesn't seems good.

**Comment by ghost** [comment]
Indeed. I move my channel to spanish speakers. And on the description I mention than that game runs at 60 fps so my low power cores cant record it at that frame rate. The game runs full speed even on 1080p.

**Comment by ghost** [comment]
back on mono games, did you tested terraria? its giving problems to me. about Dust, I have an 2.0.0.2 gog version. dont know if that supports fna already. 

hope you get kam_remake running on both platforms!

**Comment by ghost** [comment]
HI @ptitSeb , Ive received my rpi4 and I am not really happy with it. the mesa development sucks, right now master mesa runs on software, that kind of sucks. obviously there is a problem with some push.

but I forget about trying to update it (and trying virgl, which seems a good idea for qemu).
just trying bastion here... and Ive got these:
```
System.EntryPointNotFoundException: MOJOSHADER_glProgramViewportFlip
```
extremly weird, because ive recompile mojoshader and viewportflip its on... dont know how much to said about this board... I think than I will need your wrapper this time ahhahaha
maybe here it works, but as you may know, it doesnt like anholt mesa dtoverlay. but if this keeps going this bad I will give it a shot, even disabling the mesa dtoverlay.

**Comment by ghost** [comment]
@ptitSeb now its fixed... it was both... a mesa problem and a mojoshader problem!

I forget on the video to put you in the credits, but I add it on the desc!

https://www.youtube.com/watch?v=qLAiaVDMvXs&feature=youtu.be

**Comment by ptitSeb** [comment]
Ah, it works nicely on the Pi4 indeed :)

You didn't used my gl wrapper (no need with a full working GL driver anyway), but  guess you did used my fakemodex wrapper :p

**Comment by ghost** [comment]
Ohhh indeeed. It took me a lot of time to figure where was that fakemodex. I forget it was made by you!

**Comment by ptitSeb** [comment]
Yep, everything you need to make x86 code runs on Arm ;)

**Comment by ghost** [comment]
@ptitSeb do you ever tried to run any unity game on linux arm. I know its isnt supported, but maybe you find the way to do that. Id like to try unity daggerfall on the future

**Comment by ptitSeb** [comment]
Yeah, I had a look. For now, the only solution I see is using box86, but not yet, Unity is not supported. To detail a bit, issue is with the x86 libs that will be called from the custom build of mono used in unity game. There is a ticket open on Unity in Box86 repo IIRC.

**Comment by ghost** [comment]
Hope we get support on the future... I dont get why its so difficult for unity devs being arm supported on android. I know its everything about of money... but with dex and the evolved linux arm status, they should do something about that. We exist unityy guys! Hhahha, thanks! I was thinking on try your magic box86, but many linux games work around steam and that wouldnt be possible.

**Comment by ghost** [comment]
Sorry @ptitSeb , do you know the email of meverick? I have soume trouble with one of his RVGL binaries (and its closed source) and Id like to ask him some help about that. Thanks

**Comment by meveric** [comment]
What's the issue you have?
Also it's probably best if you contact me directly on the ODROID forum.

**Comment by ghost** [comment]
Ok, ive sent you a private message!
